### PR TITLE
fix: HostIterator crash

### LINKED ISSUE #705: Crash in tryLaunch / hostIterator
**Describe the bug 🐛**
In our crash reports we can see a crash caused by AlgoliaSearchClient in HTTPRequest.swift in `tryLaunch` more concretely this line: `guard let host = hostIterator.next()` [https://github.com/algolia/algoliasearch-client-swift/blob/65d8343d043c1cd9b8a0d985f325dad75c3a6805/Sources/AlgoliaSearchClient/Transport/HTTP/HTTPRequest.swift#L77](here in the repository)

**To Reproduce 🔍**
SInce it is coming through the Appstore crash reports, we could not manage to reproduce it

**Expected behavior 💭**
Should not crash

**Screenshots 🖥**

**Environment:**
 iOS 14.2 iPhone XS

### PR DESCRIPTION
**Pull request by VladislavFitz**
**Summary**

This fix adds synchronisation to HostIterator callback provided by retry strategy to avoid crash due to concurrent access to the list of hosts
Fixes #705

### CODE REVIEW

File: Sources/AlgoliaSearchClient/Transport/RetryStrategy/AlgoliaRetryStrategy.swift [hunk 1]
```diff
@@ -37,7 +37,10 @@ class AlgoliaRetryStrategy: RetryStrategy {
       }
 
       return HostIterator { [weak self] in
-        self?.hosts.first { $0.supports(callType) && $0.isUp }
+        guard let retryStrategy = self else { return nil }
```
Comment (Ant-hem) [inline Sources/AlgoliaSearchClient/Transport/RetryStrategy/AlgoliaRetryStrategy.swift:40]: Question: If I understood correctly `guard` is the `mutex/lock` of Swift, if it works like inc#/c++ shouldn't it be at the instance level to protect the instance of multiple threads?

Comment (VladislavFitz) [reply Sources/AlgoliaSearchClient/Transport/RetryStrategy/AlgoliaRetryStrategy.swift:40]: `guard` is the statement, ensuring the fulfilment of a condition in the scope bellow. 
```swift
let condition: Bool

if condition {
 // here the condition is true 
} else {
 // here the condition is false
}

guard condition else {
  // here the condition is false
}
// here the condition is true
``` 

In other words, it "protects" the code bellow from non-fulfilment of a logical expression. 
It works the same way with optional variables, unwrapping them for the code bellow. 
Here the `guard` statement ensures the code bellow that weakly captured optional `self` (equivalent of `this` in Java) is not `nil`. 

`retryStrategy.queue` is the queue used as a `lock` for access to `retryStrategy.hosts` list. It is actually present at the instance level of `AlgoliaRetryStrategy`.

Comment (Ant-hem) [reply Sources/AlgoliaSearchClient/Transport/RetryStrategy/AlgoliaRetryStrategy.swift:40]: Thanks for the explanation 👍🏻

**Review by Ant-hem — APPROVED** [review]
LGTM

**Pull request merged**